core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java

```diff
@@ -13,86 +13,186 @@
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.stream.Collectors;
 
 import static java.util.Objects.requireNonNull;
 
 public final class ConnectorExpressionTranslator
 {
+    public static final String LIKE_FUNCTION_NAME = "$like";
+    public static final String REGEXP_LIKE_FUNCTION_NAME = "regexp_like";
+
     private ConnectorExpressionTranslator() {}
 
-    public static Expression translate(ConnectorExpression expression, Map<String, Symbol> variableMappings, LiteralEncoder literalEncoder)
+    public static Expression translate(Session session, ConnectorExpression expression, Metadata metadata, LiteralEncoder literalEncoder, Optional<Map<String, Symbol>> variableMappings)
     {
-        return new ConnectorToSqlExpressionTranslator(variableMappings, literalEncoder).translate(expression);
+        return new ConnectorToSqlExpressionTranslator(session, metadata, literalEncoder, variableMappings)
+                .process(expression)
+                .orElseThrow(() -> new UnsupportedOperationException("Expression type not supported: " + expression.getClass().getName()));
     }
 
-    public static Optional<ConnectorExpression> translate(Session session, Expression expression, TypeAnalyzer types, TypeProvider inputTypes)
+    public static Optional<ConnectorExpression> translate(Session session, Expression expression, TypeAnalyzer types, TypeProvider inputTypes, Metadata metadata)
     {
-        return new SqlToConnectorExpressionTranslator(types.getTypes(session, inputTypes, expression))
+        return new SqlToConnectorExpressionTranslator(session, types.getTypes(session, inputTypes, expression), metadata)
                 .process(expression);
     }
 
     private static class ConnectorToSqlExpressionTranslator
     {
-        private final Map<String, Symbol> variableMappings;
+        private final Session session;
+        private final Metadata metadata;
         private final LiteralEncoder literalEncoder;
+        private final Optional<Map<String, Symbol>> variableMappings;
 
-        public ConnectorToSqlExpressionTranslator(Map<String, Symbol> variableMappings, LiteralEncoder literalEncoder)
+        public ConnectorToSqlExpressionTranslator(Session session, Metadata metadata, LiteralEncoder literalEncoder, Optional<Map<String, Symbol>> variableMappings)
         {
-            this.variableMappings = requireNonNull(variableMappings, "variableMappings is null");
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
             this.literalEncoder = requireNonNull(literalEncoder, "literalEncoder is null");
+            this.variableMappings = requireNonNull(variableMappings, "variableMappings is null");
         }
 
-        public Expression translate(ConnectorExpression expression)
+        public Optional<Expression> process(ConnectorExpression expression)
```

Comment:
Why did the name change?

Reply:
Please keep the structure of the method similar to how it was before (ordering of cases, non-chained `if` blocks) and just add the new cases relevant to this PR. It will keep the diff cleaner and make it easier to spot the essence of the change.

Reply:
Done